File: 105. Construct Binary Tree from Preorder and Inorder Traversal/sol.py


class TreeNode:
    def __init__(self, val: int = 0, left: 'Optional[TreeNode]' = None, right: 'Optional[TreeNode]' = None):
        self.val = val 
        self.left = left 
        self.right = right 

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if len(preorder) == 0 :
            return None
        root = TreeNode(preorder[0])
        leftLen = inorder.index(preorder[0])
        root.left = self.buildTree(preorder[1: leftLen + 1], inorder[:leftLen])
        root.right = self.buildTree(preorder[leftLen + 1: ], inorder[leftLen + 1:])
        return root

File: 105. Construct Binary Tree from Preorder and Inorder Traversal/sol.go
package constructBinaryTreeFromPreorderAndInorderTraversal

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func buildTree(preorder []int, inorder []int) *TreeNode {
	if len(preorder) == 0 {
		return nil
	}
	root := TreeNode{Val: preorder[0]}
	leftLen := 0
	for i := range preorder {
		if preorder[0] == inorder[i] {
			leftLen = i
			break
		}
	}
	root.Left = buildTree(preorder[1:leftLen+1], inorder[:leftLen])
	root.Right = buildTree(preorder[leftLen+1:], inorder[leftLen+1:])
	return &root
}
